procedure/src/main/java/org/apache/iotdb/procedure/store/ProcedureWAL.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iotdb.procedure.store;
+
+import org.apache.iotdb.procedure.Procedure;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.*;
+import java.nio.ByteBuffer;
+import java.nio.channels.FileChannel;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.List;
+
+public class ProcedureWAL {
+
+  private static final Logger LOG = LoggerFactory.getLogger(ProcedureWAL.class);
+
+  private static final String TMP_SUFFIX = ".tmp";
+  private Path walFilePath;
+  private ByteBuffer byteBuffer;
+
+  public ByteBuffer getByteBuffer() {
+    return byteBuffer;
+  }
+
+  public ProcedureWAL(Path walFilePath, ByteBuffer byteBuffer) {
+    this.walFilePath = walFilePath;
+    this.byteBuffer = byteBuffer;
+  }
+
+  /**
+   * Create a wal file
+   *
+   * @throws IOException ioe
+   */
+  public void save() throws IOException {
+    File walTmp = new File(walFilePath + TMP_SUFFIX);
+    Path walTmpPath = walTmp.toPath();
+    Files.deleteIfExists(walTmpPath);
+    Files.createFile(walTmpPath);
+    try (FileOutputStream fos = new FileOutputStream(walTmp);
+        FileChannel channel = fos.getChannel()) {
+      byteBuffer.flip();
+      channel.write(byteBuffer);
+    }
+    Files.deleteIfExists(walTmpPath);
+    Files.move(walTmpPath, walFilePath);
+  }
+
+  /**
+   * Load wal files into memory
+   *
+   * @param procedureList procedure list
+   */
+  public void load(List<Procedure> procedureList) {
+    if (procedureList == null || procedureList.isEmpty()) {
+      LOG.info("None  procedure  wal  found.");
+      return;
+    }
+    Procedure procedure = null;
+    try (FileInputStream fis = new FileInputStream(walFilePath.toFile());
+        FileChannel channel = fis.getChannel()) {
+      while (channel.read(byteBuffer) > 0) {
+        byteBuffer.flip();
+        procedure = Procedure.newInstance(byteBuffer);
```

Comment:
fixed. Add a loop for bytebuffer remaining check.

Reply:
the new loop is not enough. If the bytebuffer could contains two or more procedures, the loop could not  deserialize them corretly.  And please add a test for deserialize and serialize for the procedurewal

Reply:
Now one procedure maps to one ProcedureWAL, which contains a buffer and a filename. It won't contain two or  more procs in one bytebuffer. In order to control memory overhead, a constraint configuration will be introduced to control concurrent procedures. Is this design acceptable?

Reply:
a procedure is a file? If we delete a batch of timeseries, every timeseries will have a file?  Does it cause too many files? How do you think? @wangchao316

Reply:
a procedure is a file.  delete timeseries does not use procedure framework. there is used for procedure framework, as so delete storage group,  add node or remove node. 
why a procedure is a file?   
Because The procedure records need to be deleted frequently. Therefore, the procedure records can be deleted easily by storing them in different files.

Reply:
ok